client/src/main/java/io/pravega/client/stream/impl/ControllerResolverFactory.java

```diff
@@ -137,50 +155,116 @@ public void start(Listener listener) {
                         .loadBalancerFactory(RoundRobinLoadBalancerFactory.getInstance())
                         .usePlaintext(true)
                         .build());
-                this.scheduledExecutor = Executors.newScheduledThreadPool(1,
-                        new ThreadFactoryBuilder().setNameFormat("fetch-controllers-%d").setDaemon(true).build());
-                this.scheduledExecutor.scheduleWithFixedDelay(
-                        this::getControllers, 0L, 120L, TimeUnit.SECONDS);
             } else {
-                // Use the bootstrapped server list as the final set of controllers.
-                final ResolvedServerInfoGroup serverInfoGroup = ResolvedServerInfoGroup.builder().addAll(
-                        this.bootstrapServers.stream().map(ResolvedServerInfo::new)
-                                .collect(Collectors.toList()))
-                        .build();
-                log.info("Updating client with controllers: {}", serverInfoGroup);
-                this.resolverUpdater.onUpdate(Collections.singletonList(serverInfoGroup), Attributes.EMPTY);
+                // If the servers comprise only of IP addresses then we only need to update the controller list once.
+                if (this.bootstrapServers.stream().allMatch(
+                        inetSocketAddress -> InetAddresses.isInetAddress(inetSocketAddress.getHostString()))) {
+                    final ResolvedServerInfoGroup serverInfoGroup = ResolvedServerInfoGroup.builder()
+                            .addAll(this.bootstrapServers.stream()
+                                    .map(address -> new ResolvedServerInfo(
+                                            new InetSocketAddress(address.getHostString(), address.getPort())))
+                                    .collect(Collectors.toList()))
+                            .build();
+                    log.info("Updating client with controllers: {}", serverInfoGroup);
+                    this.resolverUpdater.onUpdate(Collections.singletonList(serverInfoGroup), Attributes.EMPTY);
+                    started.set(true);
+                    return;
+                }
+                // else case is for controllers with DNS hostnames, which needs to be resolved periodically.
             }
+
+            // Schedule the first discovery immediately.
+            this.scheduledExecutor = Executors.newScheduledThreadPool(1,
```

Comment:
My main worry is adding technical debt, but I'm equally worried that if we do it here, it will make it harder to verify. I'm tempted to follow your suggestion of addressing it separately as it doesn't add any correctness issue.